File: JudoKitObjCExampleApp/Controllers/Main/MainViewController.m
//
//  MainViewController.m
//  JudoKitObjCExample
//
//  Copyright (c) 2016 Alternative Payments Ltd
//
//  Permission is hereby granted, free of charge, to any person obtaining a copy
//  of this software and associated documentation files (the "Software"), to deal
//  in the Software without restriction, including without limitation the rights
//  to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
//  copies of the Software, and to permit persons to whom the Software is
//  furnished to do so, subject to the following conditions:
//
//  The above copyright notice and this permission notice shall be included in all
//  copies or substantial portions of the Software.
//
//  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
//  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
//  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
//  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
//  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
//  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
//  SOFTWARE.

#import <CoreLocation/CoreLocation.h>

#import "MainViewController.h"
#import "DetailViewController.h"
#import "ExampleAppCredentials.h"
#import "JPApplePayConfiguration.h"
#import "SettingsViewController.h"
#import "HalfHeightPresentationController.h"
#import "DemoFeature.h"
#import "Settings.h"
#import "JudoKitObjC.h"

static NSString * const kCellIdentifier = @"com.judo.judopaysample.tableviewcellidentifier";
static NSString * const kConsumerReference = @"judoPay-sample-app-objc";

@interface MainViewController ()

@property (nonatomic, strong) JPAmount *amount;
@property (nonatomic, strong) JPReference *reference;
@property (nonatomic, strong) JPCardDetails *cardDetails;
@property (nonatomic, strong) JPPaymentToken *payToken;
@property (nonatomic, strong) JudoKit *judoKitSession;
@property (nonatomic, strong) JPConfiguration *configuration;
@property (nonatomic, strong) JPApplePayConfiguration *applePayConfiguration;
@property (strong, nonatomic) CLLocationManager *locationManager;
@property (strong, nonatomic) NSArray <DemoFeature *> *features;
@property (strong, nonatomic) Settings *settings;

@end

@implementation MainViewController

# pragma mark - View lifecycle

- (void)viewDidLoad {
    [self initializeJudoSDK];
    [self requestLocationPermissions];
    [super viewDidLoad];
}

#pragma mark - Setup methods

- (void)initializeJudoSDK {
    self.judoKitSession = [[JudoKit alloc] initWithToken:token secret:secret];
    self.judoKitSession.isSandboxed = YES;
}

- (void)requestLocationPermissions {
    self.locationManager = [CLLocationManager new];
    [self.locationManager requestWhenInUseAuthorization];
}

#pragma mark - SDK Features

- (void)paymentOperation {
    [self.judoKitSession invokeTransactionWithType:TransactionTypePayment
                                     configuration:self.configuration
                                        completion:^(JPResponse *response, NSError *error) {
        [self handleResponse:response];
    }];
}

- (void)preAuthOperation {
    [self.judoKitSession invokeTransactionWithType:TransactionTypePreAuth
                                     configuration:self.configuration
                                        completion:^(JPResponse *response, NSError *error) {
        [self handleResponse:response];
    }];
}

- (void)createCardTokenOperation {
    [self.judoKitSession invokeTransactionWithType:TransactionTypeRegisterCard
                                     configuration:self.configuration
                                        completion:^(JPResponse *response, NSError *error) {
        [self handleResponse:response];
    }];
}

- (void)checkCardOperation {
    [self.judoKitSession invokeTransactionWithType:TransactionTypeCheckCard
                                     configuration:self.configuration
                                        completion:^(JPResponse *response, NSError *error) {
        [self handleResponse:response];
    }];
}

- (void)saveCardOperation {
    [self.judoKitSession invokeTransactionWithType:TransactionTypeSaveCard
                                     configuration:self.configuration
                                        completion:^(JPResponse *response, NSError *error) {
        [self handleResponse:response];
    }];
}

- (void)applePayPaymentOperation {
    [self.judoKitSession invokeApplePayWithMode:TransactionModePayment
                                  configuration:self.applePayConfiguration
                                     completion:^(JPResponse *response, NSError *error) {
        [self handleResponse:response];
    }];
}

- (void)applePayPreAuthOperation {
    [self.judoKitSession invokeApplePayWithMode:TransactionModePreAuth
                                  configuration:self.applePayConfiguration
                                     completion:^(JPResponse *response, NSError *error) {
        [self handleResponse:response];
    }];
}

- (void)paymentMethodOperation {
    [self.judoKitSession invokePaymentMethodScreenWithMode:TransactionModePayment
                                             configuration:self.configuration
                                                completion:^(JPResponse *response, NSError *error) {
        [self handleResponse:response];
    }];
}

- (void)preAuthMethodOperation {
    [self.judoKitSession invokePaymentMethodScreenWithMode:TransactionModePreAuth
                                             configuration:self.configuration
                                                completion:^(JPResponse *response, NSError *error) {
        [self handleResponse:response];
    }];
}

#pragma mark - Helper methods

- (void)handleResponse:(JPResponse *)response {
    
    if (!response) {
        return;
    }
    
    JPTransactionData *transactionData = response.items.firstObject;

    if (transactionData.cardDetails) {
        self.cardDetails = transactionData.cardDetails;
        self.payToken = transactionData.paymentToken;
    }
    [self dismissViewControllerAnimated:YES completion:^{
        [self presentDetailsViewControllerWithTransactionData:transactionData];
    }];
}

- (void)presentDetailsViewControllerWithTransactionData:(JPTransactionData *)transactionData {
    DetailViewController *viewController = [[DetailViewController alloc] initWithNibName:@"DetailViewController" bundle:nil];
    viewController.transactionData = transactionData;
    [self.navigationController pushViewController:viewController animated:YES];
}

- (void)presentErrorWithMessage:(NSString *)message {
    UIAlertController *alertController = [UIAlertController alertControllerWithTitle:@"Error"
                                                                             message:message
                                                                      preferredStyle:UIAlertControllerStyleAlert];

    [alertController addAction:[UIAlertAction actionWithTitle:@"OK"
                                                        style:UIAlertActionStyleCancel
                                                      handler:nil]];

    [self presentViewController:alertController animated:YES completion:nil];
}

- (UIPresentationController *)presentationControllerForPresentedViewController:(UIViewController *)presented
                                                      presentingViewController:(UIViewController *)presenting
                                                          sourceViewController:(UIViewController *)source {
    return [[HalfHeightPresentationController alloc] initWithPresentedViewController:presented
                                                            presentingViewController:presenting];
}

#pragma mark - Lazy properties

- (JPConfiguration *)configuration {
    if (!_configuration) {
        _configuration = [[JPConfiguration alloc] initWithJudoID:judoId
                                                          amount:self.amount
                                                       reference:self.reference];
        
        _configuration.siteId = siteId;
        _configuration.cardAddress = [[JPAddress alloc] initWithLine1:@"myAddress1"
                                                                line2:@"myAddress2"
                                                                line3:@"myAddress3"
                                                                 town:@"myTown"
                                                       billingCountry:@"myCountry"
                                                             postCode:@"myPostalCode"];
        
        _configuration.paymentMethods = @[JPPaymentMethod.card, JPPaymentMethod.iDeal, JPPaymentMethod.applePay];
        _configuration.supportedCardNetworks = CardNetworkVisa | CardNetworkMasterCard | CardNetworkAMEX;
        _configuration.applePayConfiguration = self.applePayConfiguration;
    }
    return _configuration;
}

- (JPApplePayConfiguration *)applePayConfiguration {
    
    NSDecimalNumber *itemOnePrice = [NSDecimalNumber decimalNumberWithString:@"0.01"];
    NSDecimalNumber *itemTwoPrice = [NSDecimalNumber decimalNumberWithString:@"0.02"];
    NSDecimalNumber *totalPrice = [NSDecimalNumber decimalNumberWithString:@"0.03"];
    
    NSArray *items = @[[JPPaymentSummaryItem itemWithLabel:@"Item 1" amount:itemOnePrice],
                       [JPPaymentSummaryItem itemWithLabel:@"Item 2" amount:itemTwoPrice],
                       [JPPaymentSummaryItem itemWithLabel:@"Tim Apple" amount:totalPrice]];
 
    JPApplePayConfiguration *configuration = [[JPApplePayConfiguration alloc] initWithMerchantId:merchantId
                                                                                        currency:self.settings.currency
                                                                                     countryCode:@"GB"
                                                                             paymentSummaryItems:items];
    configuration.requiredShippingContactFields = ContactFieldAll;
    configuration.requiredBillingContactFields = ContactFieldAll;
    configuration.returnedContactInfo = ReturnedInfoAll;

    return configuration;
}

- (JPAmount *)amount {
    if (!_amount) {
        _amount = [[JPAmount alloc] initWithAmount:@"0.01" currency:self.settings.currency];
    }
    return _amount;
}

- (JPReference *)reference {
    if (!_reference) {
        _reference = [JPReference consumerReference:kConsumerReference];
        _reference.metaData = @{@"exampleMetaKey": @"exampleMetaValue"};
    }
    return _reference;
}

- (Settings *)settings {
    if (!_settings) {
        _settings = Settings.defaultSettings;
    }
    return _settings;
}

- (NSArray<DemoFeature *> *)features {
    if (!_features) {
        _features = DemoFeature.defaultFeatures;
    }
    return _features;
}

@end

@implementation MainViewController (TableViewDelegates)

#pragma mark - UITableViewDataSource

- (NSInteger)tableView:(nonnull UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return self.features.count;
}

- (nonnull UITableViewCell *)tableView:(nonnull UITableView *)tableView
                 cellForRowAtIndexPath:(nonnull NSIndexPath *)indexPath {

    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:kCellIdentifier forIndexPath:indexPath];
    DemoFeature *option = self.features[indexPath.row];
    cell.textLabel.text = option.title;
    cell.detailTextLabel.text = option.details;
    return cell;
}

#pragma mark - UITableViewDelegate

- (void)tableView:(nonnull UITableView *)tableView didSelectRowAtIndexPath:(nonnull NSIndexPath *)indexPath {
    [tableView deselectRowAtIndexPath:indexPath animated:YES];
    DemoFeature *feature = self.features[indexPath.row];
    switch (feature.type) {
        case DemoFeatureTypePayment:
            [self paymentOperation];
            break;

        case DemoFeatureTypePreAuth:
            [self preAuthOperation];
            break;

        case DemoFeatureTypeCreateCardToken:
            [self createCardTokenOperation];
            break;

        case DemoFeatureTypeSaveCard:
            [self saveCardOperation];
            break;

        case DemoFeatureTypeCheckCard:
            [self checkCardOperation];
            break;

        case DemoFeatureTypeApplePayPayment:
            [self applePayPaymentOperation];
            break;

        case DemoFeatureTypeApplePayPreAuth:
            [self applePayPreAuthOperation];
            break;

        case DemoFeatureTypePaymentMethods:
            [self paymentMethodOperation];
            break;

        case DemoFeatureTypePreAuthMethods:
            [self preAuthMethodOperation];
            break;
    }
}

@end

@implementation MainViewController (Settings)

- (void)settingsViewController:(SettingsViewController *)viewController
             didUpdateSettings:(Settings *)settings {
    self.settings = settings;
    self.configuration.uiConfiguration.isAVSEnabled = settings.isAVSEnabled;
    self.configuration.amount = [JPAmount amount:self.configuration.amount.amount
                                        currency:settings.currency];
}
 
- (IBAction)settingsButtonHandler:(id)sender {
    SettingsViewController *svc = [[SettingsViewController alloc] initWithSettings:self.settings];
    svc.delegate = self;
    svc.modalPresentationStyle = UIModalPresentationCustom;
    svc.transitioningDelegate = self;
    [self presentViewController:svc animated:YES completion:nil];
}

@end

File: JudoKitSwiftExampleApp/ViewControllers/MainViewController.swift
import UIKit
import JudoKitObjC

class MainViewController: UITableViewController {
    var settings = Settings(isAVSEnabled: false, currency: .GBP)
    var judoKit = JudoKit(token: token, secret: secret)
    var locationManager = CLLocationManager()
    var transactionData: JPTransactionData?
    let consumerReference = "judoPay-sample-app"
    lazy var reference = JPReference(consumerReference: consumerReference)
    
    let defaultFeatures = [
        DemoFeature(type: .payment,
                    title: "Pay with card",
                    details: "by entering card details"),
        
        DemoFeature(type: .preAuth,
                    title: "PreAuth with card",
                    details: "by entering card details"),
        
        DemoFeature(type: .createCardToken,
                    title: "Register card",
                    details: "to be stored for future transactions"),
        
        DemoFeature(type: .checkCard,
                    title: "Check card",
                    details: "to validate a card"),
        
        DemoFeature(type: .saveCard,
                    title: "Save card",
                    details: "to be stored for future transactions"),
        DemoFeature(type: .applePayPayment,
                    title: "Apple Pay payment",
                    details: "with a wallet card"),
        
        DemoFeature(type: .applePayPreAuth,
                    title: "Apple Pay preAuth",
                    details: "with a wallet card"),
        
        DemoFeature(type: .paymentMethods,
                    title: "Payment Method",
                    details: "with default payment methods"),
        
        DemoFeature(type: .preAuthMethods,
                    title: "PreAuth Methods",
                    details: "with default preauth methods")
    ]
    
    var amount: JPAmount {
        return  JPAmount("0.01", currency: settings.currency.rawValue)
    }
    
    var configuration: JPConfiguration {
        let configuration = JPConfiguration(judoID: judoId, amount: self.amount, reference: reference)
        configuration.supportedCardNetworks = [.networkVisa, .networkMasterCard, .networkAMEX]
        configuration.uiConfiguration.isAVSEnabled = settings.isAVSEnabled
        return configuration
    }
    
    var applePayConfigurations: JPApplePayConfiguration {
        let items = [JPPaymentSummaryItem(label: "item 1", amount: 0.01),
                     JPPaymentSummaryItem(label: "item 2", amount: 0.02),
                     JPPaymentSummaryItem(label: "Judo Pay", amount: 0.03)]
        let configurations = JPApplePayConfiguration(merchantId: merchantId,
                                                     currency: self.settings.currency.rawValue,
                                                     countryCode: "GB",
                                                     paymentSummaryItems: items)
        return configurations
    }
    
    override func viewDidLoad() {
        super.viewDidLoad()
        self.navigationController?.isToolbarHidden = false
        judoKit?.isSandboxed = true
        locationManager.requestWhenInUseAuthorization()
    }
    
    // MARK: Table view datasource
    override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return defaultFeatures.count
    }
    
    override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cell = tableView.dequeueReusableCell(withIdentifier: "reuseIdentifier", for: indexPath)
        let feature = defaultFeatures[indexPath.row]
        
        cell.textLabel?.text = feature.title
        cell.detailTextLabel?.text = feature.details
        
        return cell
    }
    
    override func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        tableView.deselectRow(at: indexPath, animated: true)
        
        let feature = defaultFeatures[indexPath.row]
        
        switch feature.type {
        case .payment:
            pay()
            
        case .preAuth:
            preAuth()
            
        case .createCardToken:
            createCardToken()
            
        case .saveCard:
            saveCard()
            
        case .checkCard:
            checkCard()
            
        case .applePayPayment:
            applePayPayment()
            
        case .applePayPreAuth:
            applePayPreAuth()
            
        case .paymentMethods:
            navigateToPaymentMethods()
            
        case .preAuthMethods:
            navigateToPreAuthMethods()
        }
    }
    
    // MARK: - Navigation
    override func prepare(for segue: UIStoryboardSegue, sender: Any?) {
        if let navigation = segue.destination as? UINavigationController,
            let controller = navigation.viewControllers.first as? SettingsTableViewController {
            controller.settings = self.settings
            controller.delegate = self
        }
    }
    
    // MARK: Actions
    @objc func pay() {
        self.judoKit?.invokeTransaction(with: .payment,
                                        configuration: configuration,
                                        completion: {[weak self] (response, error) in
                                            self?.handle(response: response)
        })
    }
    
    @objc func preAuth() {
        self.judoKit?.invokeTransaction(with: .preAuth,
                                        configuration: configuration,
                                        completion: {[weak self] (response, error) in
                                            self?.handle(response: response)
        })
    }
    
    @objc func createCardToken() {
        self.judoKit?.invokeTransaction(with: .registerCard,
                                        configuration: configuration,
                                        completion: {[weak self] (response, error) in
                                            self?.handle(response: response)
        })
    }
    
    @objc func saveCard() {
        self.judoKit?.invokeTransaction(with: .saveCard,
                                        configuration: configuration,
                                        completion: {[weak self] (response, error) in
                                            self?.handle(response: response)
        })
    }
    
    @objc func checkCard() {
        self.judoKit?.invokeTransaction(with: .checkCard,
                                        configuration: configuration,
                                        completion: {[weak self] (response, error) in
                                            self?.handle(response: response)
        })
    }
    
    
    @objc func applePayPayment() {
        self.judoKit?.invokeApplePay(with: .payment,
                                     configuration: applePayConfigurations,
                                     completion: {[weak self] (response, error) in
                                        self?.handle(response: response)
        })
    }
    
    @objc func applePayPreAuth() {
        self.judoKit?.invokeApplePay(with: .preAuth,
                                     configuration: applePayConfigurations,
                                     completion: {[weak self] (response, error) in
                                        self?.handle(response: response)
        })
    }
    
    @objc func navigateToPaymentMethods() {
        self.judoKit?.invokePaymentMethodScreen(with: .payment,
                                                configuration: configuration,
                                                completion: {[weak self] (response, error) in
                                                    self?.handle(response: response)
        })
    }
    
    @objc func navigateToPreAuthMethods() {
        self.judoKit?.invokePaymentMethodScreen(with: .preAuth,
                                                configuration: configuration,
                                                completion: {[weak self] (response, error) in
                                                    self?.handle(response: response)
        })
    }
    
    // MARK: - Helper methods
    
    func handle(response: JPResponse?) {
        guard let response = response else {
            return
        }
        let detailsViewController = TransactionDetailsTableViewController(title: "Payment receipt",
                                                                          response: response)
        dismiss(animated: true) {
            self.navigationController?.pushViewController(detailsViewController, animated: true)
        }
    }
    
    func presentAlert(with title: String, message: String) {
        let alertController = UIAlertController(title: title, message: message, preferredStyle: .alert)
        alertController.addAction(UIAlertAction(title: "OK", style: .cancel, handler: nil))
        present(alertController, animated: true, completion: nil)
    }
}

extension MainViewController: SettingsTableViewControllerDelegate {
    func settingsTable(viewController: SettingsTableViewController, didUpdate settings: Settings) {
        self.settings = settings
    }
}
